risc0/zkp/rust/src/field/goldilocks.rs

```diff
@@ -0,0 +1,745 @@
+
+/// ! Goldilocks field.
+/// ! Support for the base finite field modulo 2^64 - 2^32 + 1
+use crate::field::{self, Elem as FieldElem};
+
+use core::ops;
+
+use bytemuck::{Pod, Zeroable};
+
+/// The Goldilocks class is an element of the finite field F_p, where P is the
+/// prime number 2^64 - 2^32 + 1. Here we implement integer
+/// arithmetic modulo P for both Goldilocks and for a field extension of
+/// Goldilocks.
+///
+/// The `Fp` datatype is the core type of all of the operations done within the
+/// zero knowledge proofs, and is the smallest 'addressable' datatype, and the
+/// base type of which all composite types are built. In many ways, one can
+/// imagine it as the word size of a strange architecture,
+/// and its operations as wrapping operations which respect word size P.
+///
+/// The Fp class wraps all standard arithmetic operations to make finite
+/// field elements appear like ordinary numbers (which, for the most part, they
+/// are).
+#[derive(Eq, PartialEq, Clone, Copy, Debug, Pod, Zeroable)]
+#[repr(transparent)]
+pub struct Elem(u64);
+
+impl Default for Elem {
+    /// As a default, return the zero [Elem].
+    fn default() -> Self {
+        Self::ZERO
+    }
+}
+
+/// The modulus of our Goldilocks field: 2^64 - 2^32 + 1
+const P: u64 = (0u64.wrapping_sub(1) << 32) + 1;
+// TODO: Vestigial, should be improved on generic implementation
+const P_U64: u64 = P;
+
+impl field::Elem for Elem {
+    const ZERO: Self = Elem::new(0u64);
+    const ONE: Self = Elem::new(1u64);
+
+    /// Compute the multiplicative inverse of `x`, or `1 / x` in finite field
+    /// terms. Since we know by Fermat's Little Theorem that
+    /// `x ^ (P - 1) == 1 % P` for any `x != 0`,
+    /// it follows that `x * x ^ (P - 2) == 1 % P` for `x != 0`.
+    /// That is, `x ^ (P - 2)` is the multiplicative inverse of `x`.
+    /// Note that if computed this way, the *inverse* of zero comes out as zero,
+    /// which we allow because it is convenient in many cases.
+    fn inv(self) -> Self {
+        self.pow((P - 2) as usize)
+    }
+
+    /// Generate a random value within our Goldilocks field
+    fn random(rng: &mut impl rand::Rng) -> Self {
+        // The range of possible RNG-generated u64 integers includes an uneven region
+        // modulo P. We want to reject u64 values from this region because, if
+        // mapped to finite field elements (wrapped), it leads to over-selection
+        // of the wrapped values. Here, P happens to fit only once into a 64-bit
+        // space, so we accept only RNG-generated u64 values less than P.
+        let mut val: u64 = rng.gen();
+        while val >= P {
+            val = rng.gen();
+        }
+        Elem::from(val)
+    }
+}
+
+macro_rules! rou_array {
+    [$($x:literal),* $(,)?] => {
+        [$(Elem::new($x)),* ]
+    }
+}
+
+/// Maximum power of two for which we have a root of unity using Goldilocks
+/// field
+impl field::RootsOfUnity for Elem {
+    const MAX_ROU_PO2: usize = 32;
```

Comment:
To follow up on my earlier comment -- IMO pull this comment to just before the definition of `MAX_ROU_PO2`.
```suggestion
impl field::RootsOfUnity for Elem {
    /// Maximum power of two for which we have a root of unity using Goldilocks
    /// field
    const MAX_ROU_PO2: usize = 32;
```